include/exec/repeat_effect_until.hpp

```diff
@@ -0,0 +1,215 @@
+          , __rcvr_((_Receiver &&) __rcvr) {
+          __source_op_.__construct_with([&] {
+            return stdexec::connect(stdexec::on(__sched_, __source_), __receiver_t{this});
+          });
+        }
+
+        friend void tag_invoke(stdexec::start_t, __t &__self) noexcept {
+          stdexec::start(__self.__source_op_.__get());
+        }
+      };
+    };
+
+    template <class _SourceId, class _ReceiverId>
+    struct __receiver<_SourceId, _ReceiverId>::__t {
+      using __id = __receiver;
+      using _Source = stdexec::__t<_SourceId>;
+      using _Receiver = stdexec::__t<_ReceiverId>;
+      using __op_t = stdexec::__t<__operation<_SourceId, _ReceiverId>>;
+
+      __op_t *__op_;
+
+      explicit __t(__op_t *op) noexcept
+        : __op_(op) {
+        STDEXEC_ASSERT(__op_ != nullptr);
+      }
+
+      __t(__t &&__other) noexcept
+        : __op_(std::exchange(__other.__op_, {})) {
+        STDEXEC_ASSERT(__op_ != nullptr);
+      }
+
+      template <same_as<__t> _Self, convertible_to<bool> _Done>
+        requires __callable<set_value_t, _Receiver>
+              && __callable<set_error_t, _Receiver, std::exception_ptr>
+      friend void tag_invoke(set_value_t, _Self &&__self, _Done &&__done_ish) noexcept {
+        bool __done = static_cast<bool>(__done_ish); // BUGBUG potentially throwing.
+        auto *__op = __self.__op_;
+
+        // The following line causes the invalidation of __self.
+        __op->__source_op_.__destruct();
+
+        // If the sender completed with true, we're done
+        if (__done) {
+          stdexec::set_value((_Receiver &&) __op->__rcvr_);
+        } else {
+          try {
+            auto &__source_op = __op->__source_op_.__construct_with([&]() {
+              return stdexec::connect(stdexec::on(__op->__sched_, __op->__source_), __t{__op});
+            });
+            stdexec::start(__source_op);
+          } catch (...) {
+            stdexec::set_error((_Receiver &&) __op->__rcvr_, std::current_exception());
+          }
+        }
+      }
+
+      template <same_as<__t> _Self>
+        requires __callable<set_stopped_t, _Receiver>
+      friend void tag_invoke(set_stopped_t, _Self &&__self) noexcept {
+        auto *__op = __self.__op_;
+        __op->__source_op_.__destruct();
+        stdexec::set_stopped((_Receiver &&) __op->__rcvr_);
+      }
+
+      template <same_as<__t> _Self, class _Error>
+        requires __callable<set_error_t, _Receiver, _Error>
+      friend void tag_invoke(set_error_t, _Self &&__self, _Error __error) noexcept {
+        auto *__op = __self.__op_;
+        __op->__source_op_.__destruct();
+        stdexec::set_error((_Receiver &&) __op->__rcvr_, (_Error &&) __error);
+      }
+
+      friend env_of_t<const _Receiver &> tag_invoke(get_env_t, const __t &__self) noexcept(
+        __nothrow_callable<get_env_t, const _Receiver &>) {
+        return get_env(const_cast<const _Receiver &>(__self.__op_->__rcvr_));
```

Comment:
I think const cast is not necessary. Why not `as_const`?

Reply:
I don't think either is necessary here. `get_env` should automatically be treating its argument as const. (I'll double-check that.)

But preferring const_cast over `as_const` for the same reason one might prefer static_cast over `std::forward`.

Reply:
Yeah, and it doesn't look like we need a const qualification, so I've removed that